File: Loan Management System/Loan Management System/src/app/Services/loan.service.js
app.service('LoanService', ['$resource', 'jwtHelper', function ($resource, jwtHelper) {
    this.applyingLoan = {};
    this.popUpFormStatus = false;
    this.yearsToYearsMonthsDays = function (totalYears) {
        const years = Math.floor(totalYears);
        const decimalMonths = (totalYears - years) * 12;
        const months = Math.floor(decimalMonths);
        const days = Math.floor((decimalMonths - months) * 30.44);

        return { years, months, days };
    }

    this.getLoanStatus = function (loan) {
        switch (loan.status) {
            case 0:
                return "Pending";
                break;
            case 1:
                return "Approved";
                break;
            case 2:
                return "Rejected";
            default:
                return null;
        }
    }

    this.applyNewLoanClicked = function(loanDetails) {
        console.log("Loan Applied");

        this.currentApplyingLoan = loanDetails;
        this.popUpFormStatus = true;
    }
    this.apiResource = $resource('/api/loan', {}, {
        getAllLoans: {
            url: '/api/loan/getAllLoans',
            method: 'GET'
        },
        getAllApplications: {
            url: '/api/loan/getAllApplications',
            method: 'GET'
        },
        getLoansByUser: {
            url: '/api/loan/getLoansByUser',
            method: 'GET',
            params: { userId: '@userId' }
        },
        applyNewLoan: {
            url: '/api/loan/applyNewLoan',
            method: 'POST'
        },
        getAllApproved: {
            url: '/api/loan/getAllApproved',
            method: 'GET'
        },
        getAllPending: {
            url: '/api/loan/getAllPending',
            method: 'GET'
        },
        getAllRejected: {
            url: '/api/loan/getAllRejected',
            method: 'GET'
        },
        approveLoan: {
            url: '/api/loan/approveLoan',
            method: 'PATCH',
            params: { loanId: '@loanId' }
        },
        rejectLoan: {
            url: '/api/loan/rejectLoan',
            method: 'PATCH',
            params: { loanId: '@loanId' }
        },
        deleteLoan: {
            url: '/api/loan/deleteLoan',
            method: 'DELETE',
            params: { loanId: '@loanId' }
        }
    });
}]);

File: Loan Management System/Loan Management System/Services/LoanService.cs
using Loan_Management_System.DTOs;
using Loan_Management_System.Models;
using Loan_Management_System.Repository.LoanData;
using Microsoft.AspNetCore.Mvc;

namespace Loan_Management_System.Services
{
    public class LoanService : ILoanService
    {
        private readonly LoanRepository _loanRepo;
        public LoanService(LoanRepository loanRepo)
        {
            _loanRepo = loanRepo;
        }
        public async Task<List<LoanDTO>> GetAllLoans()
        {
            List<Loan> allLoans = await _loanRepo.GetAllLoans();

            List<LoanDTO> loans = allLoans.Select(l => new LoanDTO
            {
                Id = l.Id,
                Title = l.Title,
                MinCreditScore = l.MinCreditScore,
                TermLength = l.TermLength,
                ProcessingFee = l.ProcessingFee,
                LogoFilePath = l.LogoPicture.FilePath,
                MinLoanAmount = l.MinLoanAmount,
                MaxLoanAmount = l.MaxLoanAmount,
                MinInterestRate = l.MinInterestRate,
                MaxInterestRate = l.MaxInterestRate
            }).ToList();
            return loans;
        }

        public async Task<List<AppliedByUser>> GetAllApplications()
        {
            List<AppliedByUser> loans = await _loanRepo.GetAllApplications();
            return loans;
        }

        public async Task<List<AppliedByUser>> GetAllApproved()
        {
            List<AppliedByUser> loans = await _loanRepo.GetAllApproved();
            return loans;
        }
        
        public async Task<List<AppliedByUser>> GetAllPending()
        {
            List<AppliedByUser> loans = await _loanRepo.GetAllPending();
            return loans;
        }

        public async Task<List<AppliedByUser>> GetAllRejected()
        {
            List<AppliedByUser> loans = await _loanRepo.GetAllRejected();
            return loans;
        }

        public async Task<List<AppliedByUser>> GetLoansByUser(int userId)
        {
            List<AppliedByUser> loans = await _loanRepo.GetLoansByUser(userId);
            return loans;
        }

        public async Task ApplyNewLoan(AppliedLoan newLoan)
        {
            newLoan.Status = Status.pending;
            newLoan.DateApplied = DateTime.Now;
            await _loanRepo.ApplyNewLoan(newLoan);
        }
        public async Task ApproveLoan(int loanId)
        {
            await _loanRepo.ApproveLoan(loanId);
        }
        public async Task RejectLoan(int loanId)
        {
            await _loanRepo.RejectLoan(loanId);
        }
        public async Task DeleteLoan(int loanId)
        {
            await _loanRepo.DeleteLoan(loanId);
        }
    }
}
